components/server/ee/src/workspace/gitpod-server-impl.ts

```diff
@@ -1827,6 +1830,69 @@ export class GitpodServerEEImpl extends GitpodServerImpl {
         return subscription;
     }
 
+    protected async ensureIsUsageBasedFeatureFlagEnabled(user: User): Promise<void> {
+        const teams = await this.teamDB.findTeamsByUser(user.id);
+        const isUsageBasedBillingEnabled = await getExperimentsClient().getValueAsync(
+            "isUsageBasedBillingEnabled",
+            false,
+            {
+                identifier: user.id,
+                custom: {
+                    team_ids: teams.map((t) => t.id).join(","),
+                    team_names: teams.map((t) => t.name).join(","),
+                },
+            },
+        );
+        if (!isUsageBasedBillingEnabled) {
+            throw new ResponseError(ErrorCodes.PERMISSION_DENIED, "not allowed");
+        }
+    }
+
+    async getStripePublishableKey(ctx: TraceContext): Promise<string | undefined> {
+        const user = this.checkAndBlockUser("getStripePublishableKey");
+        await this.ensureIsUsageBasedFeatureFlagEnabled(user);
+        return this.config.stripeSettings?.publishableKey;
+    }
+
+    async getStripeSetupIntentClientSecret(ctx: TraceContext): Promise<string | undefined> {
+        const user = this.checkAndBlockUser("getStripeSetupIntentClientSecret");
+        await this.ensureIsUsageBasedFeatureFlagEnabled(user);
+        try {
+            const setupIntent = await this.stripeService.createSetupIntent();
+            return setupIntent.client_secret || undefined;
+        } catch (error) {
+            log.error("Failed to create Stripe SetupIntent", error);
+            throw new ResponseError(500, "Failed to create Stripe SetupIntent");
+        }
+    }
+
+    async getTeamStripeCustomerId(ctx: TraceContext, teamId: string): Promise<string | undefined> {
+        const user = this.checkAndBlockUser("getTeamStripeCustomerId");
+        await this.ensureIsUsageBasedFeatureFlagEnabled(user);
+        await this.guardTeamOperation(teamId, "update");
+        try {
+            const customer = await this.stripeService.findCustomerByTeamId(teamId);
+            return customer?.id || undefined;
```

Comment:
Also, IMO you should return a 404 error when customer.id does not exist. Returning undefined prevents you classifying your errors client side and actually having reasonable metrics on what worked and didn't. This goes for `setupIntent` also.

Reply:
Hmm, interesting point. I agree on the 404 semantics. However, returning `undefined` currently means that the customer doesn't exist yet, and is handled that way in the client (different from "the Stripe API failed"). If we switch this around to a 404, we'd need to update the client as well to handle 404 as a valid case (i.e. "the team is not yet a Stripe customer").

Reply:
I think that's correct. If it doesn't exist, then it wasn't found and we should create one (if needed). In general, we should try to avoid OK responses with No data as much as possible, especially when we know why there is no data - in this case it doesn't exist in stripe.

Reply:
(Follow-up PR/issue is fine)